README.md
# OverFit-Game
OverFit Game is a fun arcade game, partly based on the PS3 game [Critter Crunch](http://www.capybaragames.com/critter-crunch-psn/)

## Requirements
- Python 3.5+
- PyGame 1.9.6
- PyZmq 18.0.1
- SQLAlchemy 1.2.12

## Installation
Simply clone the repository to your desired destination.
```bash
git clone git@github.com:samuelmovi/overfit-game.git
```

## Tests

To execute the unit tests (from project's folder):
```bash
cd server/tests
python -m unittest discover
cd ../../
cd client/tests
python -m unittest discover

```

## Usage

To correctly execute the client (from project's folder):
```bash
cd client/src/
python start.py
```

### Welcome Screen
It has three buttons:

- Single Player: takes you straight to the game screen in single-player mode
- Online: takes you to to online settings screen
- Settings: takes you to the settings screen


### Settings screen
Here there are 2 fields to fill-up:

- Online Host: server's IP address
- Player Name: this text will appear as the player name in its own game screen, and as the opponent's name in the opponent's game screen in online-play mode.

The return key must be pressed for the input to be taken in by the program.


### Online Setting Screen
It has the same input fields as the normal settings screen, and an extra button. The return key doesn't need to be pressed in this case.

The button under the fields stars the online broker, which connects to the server, authenticates, and searches for available players [AP]. If one is found the will connect and the game will be started for both. If, otherwise, no AP is found, the client will enter a wait mode, until another client challenges the AP.

During these operations, a wait-screen with different text will be shown to the client, informing them of their state.

### Game Screen
It's composed of mainly 2 sections: the information column on the left of the screen, and the rest of the screen where the action takes place [the board].

The information column is divided into 2 equal sections:

- Top section: shows the local player's name on top and, below, the information about that player's state: score, total number of figures in screen, longest column. 
- Bottom section: it has the same elements but displaying information about the opponent.

In cases of single-player games, that bottom half remains blank.


### Confirmation screens
From any screen in the game, if you press ESC key, a confirmation screen will pop-up asking you to confirm an action. Doing this in the welcome screen would prompt a question about leaving the game. In any other screen, it would ask about returning to the main screen.

Underneath the question will be 2 buttons, one for YES and one for NO. Click accordingly.

Pressing ESC while on the confirmation screen should be interpreted as clicking NO.

## Gameplay
### The Board
For lack of a better word, I'll use this word to refer to the part of the game screen where the playing happens. We could divide this board into 2 main sections:

- Taking the top 80% of the board is a series of seven columns, delimited by white vertical lines, which is where Figures are places.
- The bottom section, free of any lines, is where the player's character moves.

### The Figures
There are 5 types of figures:

- Dots: small white circle.
- Triangles: they have a small circle-shaped hole in them.
- Squares: they have a small circle-shaped hole in them.
- Triangle-holes: squares with triangle-shaped hole in them.
- Square-holes: squares with square-shaped hole in them.

You can fit dots into triangles and squares.
You can fit triangles in triangle-holes.
You can fit squares in square-holes.

If a figure's hole is full when it tries to be fitted with another one, this will cause an "OverFit", which will cause the involved figures to explode. If the exploding overfit figure is placed above, below, to the right or left, of a figure of the same kind this one will explode as well, no matter its state.

### The Player

The player is represented by a big white circle.
It moves left and right, in the bottom of the screen, below the columns [using arrow keys]. 
It has as many positions as there are columns.
It has one action button [space bar] which does one of 2 things:

- If the player is empty it will shoot a ray to capture the lowest figure in the column right above it, removing it from said column, and changing its pown state to full (represented by a small orange circle inside the player).
- If the player is full it will shoot its ray to return the figure to the column above the player (whichever onw it is). If column is full the player will be unable to do so.


## But, what's the game really about?
The purpose of the game is to score as many points as possible as efficiently as possible.

The scoring works by awarding points based on the figures destroyed:

- Dots: 1 point
- Squares/Triangles: 2 points
- Square-hole/Triangle-holes: 3 points

When figures are fitted, its score will include the value of the fitted figure as well.

To balance things out, and so it's fun, a new row of figures is added to a player's board under certain conditions:

- If the total figure count on a player's board is less than 15
- For every 15 moves the player does
- If playing online, for every 20 points your opponent makes

## Online Multiplayer
The online gaming protocol is built using ZeroMQ (through PyZmq), for the transport layer and for authentication, using ZMQ's custom ED22519 certificates.

### Server
The server script binds to 2 ports:

- Port 5555: PULL 
- Port 5556: PUB

Server receives messages from a PULL, processes them, and publishes them.

### Client
The client connects to 2 ports:

- Port 5555: PUSH
- Port 5556: SUB

### Protocol

- CLIENT clicks button `Find match` on landing page screen and sends SERVER a message with command `FIND`
- Upon receipt SERVER checks `available_players`:
    - If none are found it adds player ID to `available_players` and responds to CLIENT with `WAIT`
    - If an available player is found SERVER send both of them a `READY` message.
- Both CLIENTS must respond (could be time sensitive failover) with `READY` client status
- SERVER responds to both after last client has acknowledged with command `PLAY`
- CLIENTS will set their status as `PLAYING` and start their match


### Packets

There are client messages, and server messages.

All of them have 3 parts.

#### Client message 

Sent by client to communicate with server:

- Sender: player ID (unique 10-character random string)
- Info (json): 
    - client status: WELCOME, AVAILABLE, READY, START, PLAYING, OVER, QUIT 
    - recipient: `SERVER` or player ID
- Payload (json): match and player data

#### Server message 

Published by server:

- Recipient: player ID 
- Info (json): 
    - sender: `SERVER` or player ID
    - command: WELCOME, WAIT, READY, PLAY, WINNER
- Payload (json): forwarded payload 


## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License
[GPLv2](https://choosealicense.com/licenses/gpl-2.0/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- figure out way to spy on test classes
- allow certs.py to create required folders if non-existing
- add logic for client trying to connect to non-existent server
- check that client in available_players aren't expired
- go to landing page when leaving online game

## 2020-07-02
### Fixed
- Missing self.board.setup() in controller during game initialization
- Misshandling of figure explosion
- Controller.reset_state now checking newly created variable self.mq.connected to avoid errors exiting single player games

### Changed
- Relocated test files
- on successful save of settings data, go to start page


## 2020-01-30
### Changed
- code improvements on models: board, column
- test code for models: board, column, broker, zmq 


## 2020-01-25
### Added
- new tests for server.py, complete but not exhaustive


## 2020-01-17
### Added
- test file for server db_controller module

### Changed
- tests for server start module
- test for player model

## 2020-01-16
### Added
- when returning to online_setup screen send QUIT message and WELCOME message to server
- add event listener while connecting
- add event listener while finding matches
- new opponent dictionary specification
    
### Changed
- send landing data by default to client not on online_players
- new formatting to message printouts with timestamp

### Fixed
- Server using outdated message protocol
- client forgot to load json data before using
- crash after pressing ESC twice on waiting screen
- setting player.online as necessary for proper match negotiating
- order of methods in loop for "find_online_mach", listener is last now
- opponent's info not displaying on board
- multiplayer online game working well

## 2020-01-15
### Fixed
- Client and server communicating correctly
- sqlalchemy backend working

### Added
- view landing page code
- method to add dummy data to db if empty
- event listener for landing page

## 2020-01-13
### Changed
- removed unused code form zmq connector modules, only relevant methods left in client and server
- some modification to README for new changes, more to do
- communication protocol

## 2010-01-12
### Changed
- name of game package to client package
- name of zmq_connector modules to zmq_client and zmq_server for differentiation
- `PullPubServer` substituted by new `Server`

### Added
- Database class for database management (`db_connector.py`)
- Added `models` module for player model


## 2019-10-23
### Fixed
- maximum recursion depth issue in controller's acquire targets method
 

## 2019-10-22
### Added
- __init__ files to test folders to automate testing
- test class for player

### Changed
- renamed game.py to start.py
- moved some controller code into the models
- timeout implementation for online broker
- improved folder structure for server sub-project
- refactored server sub-project code
- refactored controller, now only one while loop in whole class

### Broke
- fatal regression in explosion, after controller refactoring

### Fixed
- some test regressions due to folder structure change

## 2019-10-19
### Changed
- new folder structure for game source files

## 2019-10-18
### Added
- test classes for model, to be completed

### Changed
- changed game.py to allow for better MVC pattern injection
- moved test folder inside game to separate from server tests
- controller.load_external_resources() now static method



game/view/view.py
import pygame
import os
from pygame.locals import *
from . import button, input_box


FPS = 30  # frames per second to update the screen
W_WIDTH = 800  # width of the program's window, in pixels
W_HEIGHT = 600  # height in pixels


class View:
	#       R   G   B
	GREY = (100, 100, 100)
	# NAVYBLUE = ( 60, 60, 100)
	WHITE = (255, 255, 255)
	RED = (255, 0, 0)
	GREEN = (0, 255, 0)
	# BLUE = ( 0,0, 255)
	YELLOW = (255, 255, 0)
	ORANGE = (255, 128, 0)
	# PURPLE = (255,0, 255)
	# CYAN = ( 0, 255, 255)
	BLACK = (0, 0, 0)
	
	base_dir = os.getcwd()

	def __init__(self):
		self.screen = pygame.display.set_mode((W_WIDTH, W_HEIGHT))
		
		self.SMALLFONT = None
		self.MEDIUMFONT = None
		self.BIGFONT = None
		self.ball = None
		self.triangle_full = None
		self.triangle_empty = None
		self.square_full = None
		self.square_empty = None
		self.triangle_hole_complete = None
		self.triangle_hole_full = None
		self.triangle_hole_empty = None
		self.square_hole_complete = None
		self.square_hole_full = None
		self.square_hole_empty = None
		self.player_full = None
		self.player_empty = None
		self.fire1 = None
		self.fire2 = None
		self.fire3 = None
		self.ray = None
	
	def set_up_fonts(self):
		self.SMALLFONT = pygame.font.Font('freesansbold.ttf', 22)
		self.MEDIUMFONT = pygame.font.Font('freesansbold.ttf', 28)
		self.BIGFONT = pygame.font.Font('freesansbold.ttf', 34)
	
	# DISPLAY-DRAWING FUNCTIONS
	def draw_start_screen(self):
		self.screen.fill(self.BLACK)
		welcome_button = button.MyButton(self.BIGFONT, 'Welcome to WeeriMeeris', self.BLACK, self.GREEN)
		welcome_button.set_coords(W_WIDTH/2, 100)

		single_player_button = button.MyButton(self.MEDIUMFONT, 'Single Player', self.RED, self.YELLOW)
		single_player_button.set_coords(W_WIDTH/2, W_HEIGHT/2)

		online_multi_button = button.MyButton(self.MEDIUMFONT, 'Online Multiplayer', self.RED, self.YELLOW)
		online_multi_button.set_coords(W_WIDTH/2, W_HEIGHT/2 + 50)

		settings_button = button.MyButton(self.MEDIUMFONT, 'Settings', self.RED, self.ORANGE)
		settings_button.set_coords(W_WIDTH/2, W_HEIGHT/2 + 150)

		self.screen.fill(self.BLACK)
		self.screen.blit(welcome_button.surface, welcome_button.rect)
		self.screen.blit(single_player_button.surface, single_player_button.rect)
		self.screen.blit(online_multi_button.surface, online_multi_button.rect)
		self.screen.blit(settings_button.surface, settings_button.rect)

		return single_player_button, online_multi_button, settings_button

	def draw_game_board(self, player):
		self.screen.fill(self.BLACK)
		# draw boundary lines
		pygame.draw.line(self.screen, self.WHITE, (150, 0), (150, W_HEIGHT), 4)
		pygame.draw.line(self.screen, self.WHITE, (0, 300), (150, 300), 4)
		for n in range(7):
			pygame.draw.line(self.screen, self.WHITE, (150+90*n, 0), (150+90*n, W_HEIGHT-100), 4)
		# player 1 score board
		# draw player name
		player_name = button.MyButton(self.MEDIUMFONT, player.name, self.WHITE, self.GREY)
		player_name.set_coords(75, 25)
		self.screen.blit(player_name.surface, player_name.rect)
		# draw player score
		player_score_title = button.MyButton(self.SMALLFONT, 'SCORE', self.RED, self.BLACK)
		player_score_title.set_coords(75, 65)
		self.screen.blit(player_score_title.surface, player_score_title.rect)
		player_score = button.MyButton(self.BIGFONT, format(player.score), self.RED, self.BLACK)
		player_score.set_coords(75, 105)
		self.screen.blit(player_score.surface, player_score.rect)
		# draw total number of figures at play
		player_total_title = button.MyButton(self.SMALLFONT, 'TOTAL', self.GREEN, self.BLACK)
		player_total_title.set_coords(75, 145)
		self.screen.blit(player_total_title.surface, player_total_title.rect)
		player_total_title = button.MyButton(self.BIGFONT, format(player.total), self.GREEN, self.BLACK)
		player_total_title.set_coords(75, 185)
		self.screen.blit(player_total_title.surface, player_total_title.rect)
		# draw occupancy of the most occupied column
		player_longest_title = button.MyButton(self.SMALLFONT, 'LONGEST', self.YELLOW, self.BLACK)
		player_longest_title.set_coords(75, 225)
		self.screen.blit(player_longest_title.surface, player_longest_title.rect)
		player_longest_title = button.MyButton(self.BIGFONT, format(player.longest), self.YELLOW, self.BLACK)
		player_longest_title.set_coords(75, 265)
		self.screen.blit(player_longest_title.surface, player_longest_title.rect)

	def draw_online_options_screen(self, player_name, ip):
		banner = self.BIGFONT.render('   SETTINGS   ', 1, self.BLACK, self.GREEN)
		banner_rect = banner.get_rect()
		banner_rect.center = (W_WIDTH / 2, 100)

		host = self.SMALLFONT.render('Online Host', 1, self.GREEN, self.GREY)
		host_rect = host.get_rect()
		host_rect.center = (W_WIDTH / 2 - 100, 175)
		host_input = input_box.InputBox(W_WIDTH / 2 + 25, 160, 150, 30, ip)

		name = self.SMALLFONT.render('Player Name', 1, self.GREEN, self.GREY)
		name_rect = name.get_rect()
		name_rect.center = (W_WIDTH / 2 - 100, 225)
		name_input = input_box.InputBox(W_WIDTH / 2 + 25, 210, 150, 30, player_name)

		start_multi_button = button.MyButton(self.MEDIUMFONT, 'Online Multiplayer', self.RED, self.YELLOW)
		start_multi_button.set_coords(W_WIDTH / 2, W_HEIGHT / 2 + 50)

		inputs = (host_input, name_input)

		self.screen.fill(self.BLACK)
		self.screen.blit(banner, banner_rect)
		self.screen.blit(host, host_rect)
		self.screen.blit(name, name_rect)
		self.refresh_input(inputs)
		self.screen.blit(start_multi_button.surface, start_multi_button.rect)

		return inputs, start_multi_button

	def draw_settings_screen(self, player_name, ip):
		banner = self.BIGFONT.render('   SETTINGS   ', 1, self.BLACK, self.GREEN)
		banner_rect = banner.get_rect()
		banner_rect.center = (W_WIDTH/2, 100)

		host = self.SMALLFONT.render('Online Host', 1, self.GREEN, self.GREY)
		host_rect = host.get_rect()
		host_rect.center = (W_WIDTH/2-100, 175)
		host_input = input_box.InputBox(W_WIDTH/2+25, 160, 150, 30, ip)

		name = self.SMALLFONT.render('Player Name', 1, self.GREEN, self.GREY)
		name_rect = name.get_rect()
		name_rect.center = (W_WIDTH / 2 - 100, 225)
		name_input = input_box.InputBox(W_WIDTH/2+25, 210, 150, 30, player_name)

		save_button = button.MyButton(self.MEDIUMFONT, 'Save', self.RED, self.YELLOW)
		save_button.set_coords(W_WIDTH / 2, W_HEIGHT / 2 + 50)

		inputs = (host_input, name_input)

		self.screen.fill(self.BLACK)
		self.screen.blit(banner, banner_rect)
		self.screen.blit(host, host_rect)
		self.screen.blit(name, name_rect)
		self.refresh_input(inputs)
		self.screen.blit(save_button.surface, save_button.rect)

		return inputs, save_button

	def draw_wait_screen(self, text='  WAITING FOR AVAILABLE ONLINE PLAYER  '):
		banner = button.MyButton(self.BIGFONT, text, self.GREEN, self.BLACK)
		banner.set_coords(W_WIDTH/2, 200)

		dots = button.MyButton(self.BIGFONT, '...', self.GREEN, self.BLACK)
		dots.set_coords(W_WIDTH/2, 300)

		self.screen.fill(self.BLACK)
		self.screen.blit(banner.surface, banner.rect)
		self.screen.blit(dots.surface, dots.rect)

	def refresh_input(self, inputs):
		for instance in inputs:
			instance.update()
			instance.draw(self.screen)

	def draw_figures(self, columns):
		image = None
		x = 90
		for n in range(7):
			x += 90
			y = 0
			for figure in columns[n].figures:
				y += 50
				if figure.shape == 'ball':
					image = self.ball
				elif figure.shape == 'triangle':
					if figure.value == 2:
						image = self.triangle_empty
					elif figure.value == 3:
						image = self.triangle_full
				elif figure.shape == 'square':
					if figure.value == 2:
						image = self.square_empty
					elif figure.value == 3:
						image = self.square_full
				elif figure.shape == 'triangle-hole':
					if figure.value == 3:
						image = self.triangle_hole_empty
					elif figure.value == 5:
						image = self.triangle_hole_full
					elif figure.value == 6:
						image = self.triangle_hole_complete
				elif figure.shape == 'square-hole':
					if figure.value == 3:
						image = self.square_hole_empty
					elif figure.value == 5:
						image = self.square_hole_full
					elif figure.value == 6:
						image = self.square_hole_complete

				self.screen.blit(image, (x, y))

	def draw_player(self, player):
		if player.status == 'full':
			image = self.player_full
		else:
			image = self.player_empty
		self.screen.blit(image, (player.position*90+175, 525))

	def draw_opponent(self, opponent):
		# draw opponent name
		player_name = button.MyButton(self.MEDIUMFONT, opponent['name'], self.WHITE, self.GREY)
		player_name.set_coords(75, 325)
		self.screen.blit(player_name.surface, player_name.rect)
		# draw opponent score
		player_score_title = button.MyButton(self.SMALLFONT, 'SCORE', self.RED, self.BLACK)
		player_score_title.set_coords(75, 365)
		self.screen.blit(player_score_title.surface, player_score_title.rect)
		player_score = button.MyButton(self.BIGFONT, str(opponent['score']), self.RED, self.BLACK)
		player_score.set_coords(75, 405)
		self.screen.blit(player_score.surface, player_score.rect)
		# draw total number of figures at play
		player_total_title = button.MyButton(self.SMALLFONT, 'TOTAL', self.GREEN, self.BLACK)
		player_total_title.set_coords(75, 445)
		self.screen.blit(player_total_title.surface, player_total_title.rect)
		player_total_title = button.MyButton(self.BIGFONT, format(opponent['total']), self.GREEN, self.BLACK)
		player_total_title.set_coords(75, 485)
		self.screen.blit(player_total_title.surface, player_total_title.rect)
		# draw occupancy of the most occupied column
		player_longest_title = button.MyButton(self.SMALLFONT, 'LONGEST', self.YELLOW, self.BLACK)
		player_longest_title.set_coords(75, 525)
		self.screen.blit(player_longest_title.surface, player_longest_title.rect)
		player_longest_title = button.MyButton(self.BIGFONT, format(opponent['longest']), self.YELLOW, self.BLACK)
		player_longest_title.set_coords(75, 565)
		self.screen.blit(player_longest_title.surface, player_longest_title.rect)

	def animate_capture(self, ray_coords):
		if ray_coords['y'] >= ray_coords['top']:
			ray_coords['y'] = self.draw_ray(ray_coords)
			ray_coords['c'] += 1
			return True
		else:
			return False

	def animate_return(self, ray_coords):
		if ray_coords['y'] >= ray_coords['top']:
			ray_coords['y'] = self.draw_ray(ray_coords)
			ray_coords['c'] += 1
			return True
		else:
			return False

	def draw_explosion(self, targets, frame):
		image = None
		if frame < 6:
			image = self.fire1
		if frame < 11:
			image = self.fire2
		elif frame < 16:
			image = self.fire3

		for target in targets:
			position, height = target
			x = position * 90 + 175
			y = height * 50 + 50
			self.screen.blit(image, (x, y))

	def draw_ray(self, ray_coords):
		top = ray_coords['top']
		x = ray_coords['x']
		y = ray_coords['y']
		c = ray_coords['c']
		if y >= top:
			y = 500
			for i in range(c):
				y = y - 10 * c
				self.screen.blit(self.ray, (x, y))
		return y

	def draw_game_over(self):
		banner = button.MyButton(self.BIGFONT, '  GAME OVER  ', self.BLACK, self.YELLOW)
		banner.set_coords(W_WIDTH/2, W_HEIGHT/2)
		self.screen.blit(banner.surface, banner.rect)

	def draw_victory(self):
		banner = button.MyButton(self.BIGFONT, '  VICTORY!!!  ', self.BLACK, self.GREEN)
		banner.set_coords(W_WIDTH/2, W_HEIGHT/2)
		self.screen.blit(banner.surface, banner.rect)

	def confirm_exit(self):
		filter_rect = pygame.Rect((0, 0), (W_WIDTH, W_HEIGHT))
		pygame.draw.rect(self.screen, self.GREY, filter_rect)

		title = button.MyButton(self.MEDIUMFONT, 'CONFIRM EXIT', self.RED, self.BLACK)
		title.set_coords(W_WIDTH/2, 200)

		yes_button = button.MyButton(self.SMALLFONT, 'YES', self.BLACK, self.RED)
		yes_button.set_coords(W_WIDTH/2-100, 400)
		no_button = button.MyButton(self.SMALLFONT, 'NO', self.BLACK, self.GREEN)
		no_button.set_coords(W_WIDTH/2+100, 400)

		self.screen.fill(self.BLACK)
		self.screen.blit(yes_button.surface, yes_button.rect)
		self.screen.blit(no_button.surface, no_button.rect)
		self.screen.blit(title.surface, title.rect)

		return yes_button, no_button

	def confirm_leave_game(self):
		filter_rect = pygame.Rect((0, 0), (W_WIDTH, W_HEIGHT))
		pygame.draw.rect(self.screen, self.GREY, filter_rect)

		title = button.MyButton(self.MEDIUMFONT, 'CONFIRM LEAVE GAME', self.RED, self.BLACK)
		title.set_coords(W_WIDTH/2, 200)

		yes_button = button.MyButton(self.SMALLFONT, 'YES', self.BLACK, self.RED)
		yes_button.set_coords(W_WIDTH/2-100, 400)
		no_button = button.MyButton(self.SMALLFONT, 'NO', self.BLACK, self.GREEN)
		no_button.set_coords(W_WIDTH/2+100, 400)

		self.screen.fill(self.BLACK)
		self.screen.blit(title.surface, title.rect)
		self.screen.blit(yes_button.surface, yes_button.rect)
		self.screen.blit(no_button.surface, no_button.rect)

		return yes_button, no_button

	def update_game_screen(self, columns, player1):
		self.draw_game_board(player1)
		self.draw_figures(columns)
		self.draw_player(player1)


if __name__ == '__main__':
	pass
